libs/dmlf/src/basic_vm_engine.cpp

```diff
@@ -389,117 +196,267 @@ bool BasicVmEngine::HasState(std::string const &name) const
+      break;
+    }
+    case fetch::vm::TypeIds::Float64:
+    {
+      output = vmVariant.Get<double>();
+      break;
+    }
+    case fetch::vm::TypeIds::Fixed32:
+    {
+      output = vmVariant.Get<fp32_t>();
+      break;
+    }
+    case fetch::vm::TypeIds::Fixed64:
+    {
+      output = vmVariant.Get<fp64_t>();
+      break;
+    }
+    case vm::TypeIds::Void:
+    case vm::TypeIds::Unknown:
+    {
+      break;
+    }
+    default:
+      return serializationError("Could not transform primitive type " +
+                                vm.GetTypeName(vmVariant.type_id));
+    }
   }
-  case fetch::vm::TypeIds::Float64:
+  else if (vmVariant.type_id == vm::TypeIds::String)
   {
-    return VmVariant(ledgerVariant.As<double>(), typeId);
+    output = vmVariant.Get<vm::Ptr<vm::String>>()->str;
   }
-  case fetch::vm::TypeIds::Fixed32:
-  case fetch::vm::TypeIds::Fixed64:
+  else
   {
-    return VmVariant(ledgerVariant.As<fp64_t>(), typeId);
-  }
-  default:
-    return VmVariant();
+    ExecutionContext executionContext(&vm, exec);
+
+    auto inside = vmVariant.Get<vm::Ptr<vm::Object>>();
+
+    serializers::MsgPackSerializer serializer;
+    try
+    {
+      inside->SerializeTo(serializer);
+      serializer.seek(0);
+    }
+    catch (std::exception &ex)
+    {
+      return serializationError("Serializing output threw error " + std::string(ex.what()));
+    }
+    catch (...)
+    {
+      return serializationError("Serializing output threw error. No details.");
+    }
+
+    try
+    {
+      serializer >> output;
+    }
+    catch (std::exception &ex)
+    {
+      return serializationError(" Deserializing output after running. Threw error " +
+                                std::string(ex.what()));
+    }
+    catch (...)
+    {
+      return serializationError(" Deserializing output after running. No details.");
+    }
+
+    if (output.IsArray())  // Convert inner type from int to fixedpoint if necessary
+    {
+      auto GetInnermostTypeId = [&vm](LedgerVariant const &ledgerVariant,
+                                      VmVariant const &    vmVariant) {
+        int                  depth   = 0;
+        LedgerVariant const *current = &ledgerVariant;
+        while (current->IsArray())
+        {
+          ++depth;
+          current = &(*current)[0];
+        }
+
+        auto result = vmVariant.type_id;
+        if (depth >= 1)
+        {
+          auto currentTypeId = vmVariant.type_id;
+          for (int i = 0; i < depth - 1; ++i)
+          {
+            currentTypeId = vm.GetTypeInfo(currentTypeId).template_parameter_type_ids[0];
+          }
+          result = vm.GetTypeInfo(currentTypeId).template_parameter_type_ids[0];
+        }
+
+        return result;
+      };
+
+      auto const innermostType = GetInnermostTypeId(output, vmVariant);
+
+      if (innermostType == vm::TypeIds::Fixed64)
+      {
+        std::function<void(LedgerVariant &)> toFixed64;
+        toFixed64 = [&toFixed64](LedgerVariant &var) {
```

Comment:
If a local function needs reference to itself for recursion, does it make sense to define it in the namespace (and `LedgerVariant` alongside)?

Reply:
I am not sure what you are saying here. Could you clarify?

Reply:
Oh, sorry, disregard `LedgerVariant` in comment. I mean, perhaps it makes sense to define `ToFixed64()` outside of `PrepOutput()`, to get rid of `std::function` object only instantiated to be recursive. :)

Reply:
Oh, I see. Mmmm well it is just to get around the lambda not being able to defined as recursive.  I don't think it is too costly and I think it nicely keeps things local.

Reply:
Actually I think it is better if we can save it. I'll update that